Gems/Terrain/Assets/Shaders/Terrain/TerrainCommon.azsli

```diff
@@ -33,14 +21,131 @@ ShaderResourceGroup ObjectSrg : SRG_PerObject
     };
 
     TerrainData m_terrainData;
+
+    // The below shouldn't be in this SRG but needs to be for now because the lighting functions depend on them.
+
+    //! Reflection Probe (smallest probe volume that overlaps the object position)
+    struct ReflectionProbeData
+    {
+        row_major float3x4 m_modelToWorld;
+        row_major float3x4 m_modelToWorldInverse; // does not include extents
+        float3 m_outerObbHalfLengths;
+        float3 m_innerObbHalfLengths;
+        float m_padding;
+        bool m_useReflectionProbe;
+        bool m_useParallaxCorrection;
+    };
+
+    ReflectionProbeData m_reflectionProbeData;
+    TextureCube m_reflectionProbeCubeMap;
+    
+    float4x4 GetReflectionProbeWorldMatrix()
+    {
+        float4x4 modelToWorld = float4x4(
+            float4(1, 0, 0, 0),
+            float4(0, 1, 0, 0),
+            float4(0, 0, 1, 0),
+            float4(0, 0, 0, 1));
+
+        modelToWorld[0] = m_reflectionProbeData.m_modelToWorld[0];
+        modelToWorld[1] = m_reflectionProbeData.m_modelToWorld[1];
+        modelToWorld[2] = m_reflectionProbeData.m_modelToWorld[2];
+        return modelToWorld;
+    }
+
+    float4x4 GetReflectionProbeWorldMatrixInverse()
+    {
+        float4x4 modelToWorldInverse = float4x4(
+            float4(1, 0, 0, 0),
+            float4(0, 1, 0, 0),
+            float4(0, 0, 1, 0),
+            float4(0, 0, 0, 1));
+
+        modelToWorldInverse[0] = m_reflectionProbeData.m_modelToWorldInverse[0];
+        modelToWorldInverse[1] = m_reflectionProbeData.m_modelToWorldInverse[1];
+        modelToWorldInverse[2] = m_reflectionProbeData.m_modelToWorldInverse[2];
+        return modelToWorldInverse;
+    }
+}
+
+ShaderResourceGroup TerrainMaterialSrg : SRG_PerMaterial
+{
+    Texture2D<float> m_heightmapImage;
+
+    Sampler HeightmapSampler
+    {
+        MinFilter = Linear;
+        MagFilter = Linear;
+        MipFilter = Point;
+        AddressU = Clamp;
+        AddressV = Clamp;
+        AddressW = Clamp;
+    };
+
+    float3 m_baseColor;
+    float m_roughness;
 }
 
+option bool o_useTerrainSmoothing = false;
+
 struct VertexInput
 {
     float2 m_position : POSITION;
     float2 m_uv : UV;
 };
 
+// Sample a texture with a 5 tap B-Spline. Consider ripping this out and putting in a more general location.
+// This function samples a 4x4 neighborhood around the uv. Normally this would take 16 samples, but by taking
+// advantage of bilinear filtering this can be done with 9 taps on the edges between pixels. The cost is further
+// reduced by dropping the diagonals.
+float SampleBSpline5Tap(Texture2D<float> texture, SamplerState textureSampler, float2 uv, float2 textureSize, float2 rcpTextureSize)
```

Comment:
Yeah I actually mentioned that in the comment for the function, but I'd like there to be a better method to avoid code duplication... the problem is right now we'd need separate functions depending on how many channels you need - this just uses r for grayscale, but you might want rgb or rgba... maybe even rg for normals. Would be great if that could be templated. The other thing is weights should really be separated from the sampling. Anyway there's code similar to this in TAA and Lut Blending. It really should be unified at some point as you say.